public/app/core/utils/explore.ts

```diff
@@ -72,20 +72,36 @@ export async function getExploreUrl(args: GetExploreUrlArguments): Promise<strin
 
   if (exploreDatasource) {
     let state: Partial<ExploreUrlState> = { range: toURLRange(timeRange.raw) };
-    if (exploreDatasource.interpolateVariablesInQueries) {
-      state = {
-        ...state,
-        datasource: exploreDatasource.uid,
-        queries: exploreDatasource.interpolateVariablesInQueries(exploreTargets, scopedVars ?? {}),
-      };
+    if (!exploreDatasource.meta.mixed) {
+      if (exploreDatasource.interpolateVariablesInQueries) {
+        state = {
+          ...state,
+          datasource: exploreDatasource.uid,
+          queries: exploreDatasource.interpolateVariablesInQueries(exploreTargets, scopedVars ?? {}),
+        };
+      } else {
+        state = {
+          ...state,
+          datasource: exploreDatasource.uid,
+          queries: exploreTargets,
+        };
+      }
     } else {
+      let queriesClone = [...queries];
```

Comment:
One possibility to consider is to not branch based on if the root DS is mixed - always get the DS from the query and only use the root one if the query one is not defined. I didn't go this route to minimize impact, but let me know if I should change this.

Reply:
It won't work with `forEach...` - state will be updated before async calls are finished. We would need to use `Promise.all` here.

> always get the DS from the query

According to the type definition uid in the ref may not be defined 🤔  is it defined when going from dashboards?

Reply:
@ifrost It will always be defined if the root datasource is mixed. If root is mixed and there is no query datasource defined we have no idea what datasource to use. 

If we decide we want to remove the branch where we only do this logic if the root datasource is mixed, and instead we want to loop through all the queries and refer to the root datasource if the query datasource is not defined, that would also solve this problem. I implemented it this way to minimize the impact the code changes would have. 

I will change the forEach to use Promise.all

Reply:
I think having a single branch is better. 

However, I understand your concern that it may change the behavior in an obscure case when there's a mismatch between the root data source and the query data source. The current logic is that we override query data source with the root anyway: https://github.com/grafana/grafana/blob/ab451d76ac6e2f3949879905ce1a53ef83d59dfc/public/app/features/explore/hooks/useStateSync/index.ts#L180 so for non-mixed uses case should not matter and we will have simpler code in here. Or is there any other use case?

Reply:
i've been looking at this a bit and i think it'd probably be better to do some cleanup anyway as the the logic was a bit obscure, and i agree that removing that branching would simplify it a lot.
i took the liberty of writing an implementation that should align a bit better with what the explore initialization logic in the state sync hook does, wdyt?

```ts
const interpolatedQueries = (
    await Promise.allSettled(
      queries
        // Explore does not support expressions so filter those out
        .filter((q) => q.datasource?.uid !== ExpressionDatasourceUID)
        .map(async (q) => {
          // if the query defines a datasource, use that one, otherwise use the one from the panel, which should always be defined.
          // this will rejects if the datasource is not found, or return the default one if dsRef is not provided.
          const queryDs = await getDataSourceSrv().get(q.datasource || dsRef);

          return {
            // interpolate the query using its datasource `interpolateVariablesInQueries` method if defined, othewise return the query as-is.
            ...(queryDs.interpolateVariablesInQueries?.([q], scopedVars ?? {})[0] || q),
            // But always set the datasource as it's  required in Explore.
            // NOTE: if for some reason the query has the "mixed" datasource, we omit the property;
            // Upon initialization, Explore use its own logic to determine the datasource.
            ...(!queryDs.meta.mixed && { datasource: queryDs.getRef() }),
          };
        })
    )
  )
    // filter out queries for which the datasource was not found
    .filter(isFulfilled)
    .map((q) => q.value);

  const exploreState = JSON.stringify({
    [generateExploreId()]: { range: toURLRange(timeRange.raw), queries: interpolatedQueries, datasource: dsRef?.uid },
  });
  return urlUtil.renderUrl('/explore', { panes: exploreState, schemaVersion: 1 });
```

Reply:
I like how compact this is, but the nice thing with the previous implementation is if there was some sort of invalidity between the root/query datasources, it would return undefined and the link would just not navigate anywhere. This seems to me like a more graceful and obvious broken state than navigating to explore and getting re-routed back to the default datasource. I overall like this solution though so I committed it.

Reply:
mmmh, good point.
however i think in this case it'd be better to just not display the link, like we do if the panel uses the shared dashboard query:

```ts
  if (
    contextSrv.hasAccessToExplore() &&
    !(panel.plugin && panel.plugin.meta.skipDataQuery) &&
    panel.datasource?.uid !== SHARED_DASHBOARD_QUERY
  ) {
    menu.push({
      text: t('panel.header-menu.explore', `Explore`),
      iconClassName: 'compass',
      onClick: onNavigateToExplore,
      shortcut: 'p x',
    });
  }
```

i'd say is not `getExploreUrl` function's responsibility to decide whether Dashboard should display the link or not, perhaps we could do some validation there (not in this PR). what do you think?